### PR TITLE
Use DependencyFactory and ConfigToolDependency for MPI

### PR DESCRIPTION
**Pull request by dcbaker**
The main code savings here is the use of the ConfigTool dependency, but this brings the MPI code in line with the way most other dependencies work, plus the code savings is nice.

### CODE REVIEW

File: mesonbuild/dependencies/mpi.py [hunk 1]
```diff
@@ -12,112 +12,102 @@
...
-
-            if not self.is_found and mesonlib.is_windows():
-                # only Intel Fortran compiler is compatible with Microsoft MPI at this time.
-                if language == 'fortran' and cid != 'intel-cl':
-                    return
-                result = self._try_msmpi()
-                if result is not None:
-                    self.is_found = True
-                    self.version, self.compile_args, self.link_args = result
+            # OpenMPI, which doesn't work with intel
```
Comment (scivision) [inline mesonbuild/dependencies/mpi.py]: https://www.open-mpi.org/faq/?category=building#build-intel-compilers

OpenMPI can indeed work with Intel compilers. I think I made this incorrect comment that OpenMPI doesn't work with Intel. However, this is not necessary to fix now.

**Comment by scivision** [comment]
Windows 10 1903:

```
>meson build
The Meson build system
Version: 0.53.999
Source dir: C:\Users\micha\code\meson\test cases\frameworks\17 mpi
Build dir: C:\Users\micha\code\meson\test cases\frameworks\17 mpi\build
Build type: native build
Project name: mpi
Project version: undefined
C compiler for the host machine: icl (intel-cl 19.1.0.166)
C linker for the host machine: xilink.exe xilink 19.1.0.166
C++ compiler for the host machine: icl (intel-cl 19.1.0.166)
C++ linker for the host machine: xilink.exe xilink 19.1.0.166
Host machine cpu family: x86_64
Host machine cpu: x86_64
Did not find pkg-config by name 'pkg-config'
Found Pkg-config: NO
mpicc found: NO
Run-time dependency MPI for c found: NO (tried pkgconfig and config-tool)

meson.build:8:2: ERROR: Problem encountered: MESON_SKIP_TEST: MPI not found, skipping.
```

**Comment by scivision** [comment]
Ubuntu 18.04 & ifort

```
$ meson build
The Meson build system
Version: 0.53.999
Source dir: meson/test cases/frameworks/17 mpi
Build dir: meson/test cases/frameworks/17 mpi/build
Build type: native build
Project name: mpi
Project version: undefined
C compiler for the host machine: icc (intel 19.1.0.166 "icc (ICC) 19.1.0.166 20191121")
C linker for the host machine: icc GNU ld.bfd 2.30
C++ compiler for the host machine: icpc (intel 19.1.0.166 "icpc (ICC) 19.1.0.166 20191121")
C++ linker for the host machine: icpc GNU ld.bfd 2.30
Host machine cpu family: x86_64
Host machine cpu: x86_64
mpiicc found: YES (/home/box/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiicc)
Run-time dependency MPI for c found: YES
mpiicpc found: YES (/home/box/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiicpc)
Run-time dependency MPI for cpp found: YES
Checking if "C++ MPI" with dependency mpiicpc links: NO
Fortran compiler for the host machine: ifort (intel 19.1.0.166 "ifort (IFORT) 19.1.0.166 20191121")
Fortran linker for the host machine: ifort GNU ld.bfd 2.30
mpiifort found: YES (/home/box/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort)
Run-time dependency MPI for fortran found: YES
Checking if "Fortran MPI" with dependency mpiifort links: NO
Build targets in project: 1

Found ninja-1.10.0 at /home/box/.local/bin/ninja


$ meson test -C build

ninja: Entering directory `meson/test cases/frameworks/17 mpi/build'
[1/2] Compiling C object 'exec@exe/main.c.o'.
FAILED: exec@exe/main.c.o
icc -Iexec@exe -I. -I.. -pipe -D_FILE_OFFSET_BITS=64 -Wall -w3 -diag-disable:remark -g -traceback -O0 -g -MD -MQ 'exec@exe/main.c.o' -MF 'exec@exe/main.c.o.d' -o 'exec@exe/main.c.o' -c ../main.c
../main.c(2): catastrophic error: cannot open source file "mpi.h"
  #include <mpi.h>
                  ^

compilation aborted for ../main.c (code 4)
ninja: build stopped: subcommand failed.
Could not rebuild meson/test cases/frameworks/17 mpi/build
```

**Comment by scivision** [comment]
Ubuntu 18.04 & gfortran   seems OK

```
meson build
The Meson build system
Version: 0.53.999
Source dir: meson/test cases/frameworks/17 mpi
Build dir: meson/test cases/frameworks/17 mpi/build
Build type: native build
Project name: mpi
Project version: undefined
C compiler for the host machine: gcc (gcc 7.4.0 "gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
C linker for the host machine: gcc GNU ld.bfd 2.30
C++ compiler for the host machine: g++ (gcc 7.4.0 "g++ (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
C++ linker for the host machine: g++ GNU ld.bfd 2.30
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Run-time dependency MPI for c found: YES 2.1.1
Run-time dependency MPI for cpp found: YES 2.1.1
Checking if "C++ MPI" with dependency ompi-cxx links: YES
Fortran compiler for the host machine: gfortran (gcc 7.4.0 "GNU Fortran (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
Fortran linker for the host machine: gfortran GNU ld.bfd 2.30
Run-time dependency MPI for fortran found: YES 2.1.1
Checking if "Fortran MPI" with dependency ompi-fort links: YES
Run-time dependency MPI found: YES 2.1.1
Build targets in project: 3

Found ninja-1.10.0 at .local/bin/ninja


$ meson test -C build
ninja: Entering directory `meson/test cases/frameworks/17 mpi/build'
[6/6] Linking target execpp.
1/3 MPI C                                   OK       0.39 s
2/3 MPI C++                                 OK       0.39 s
3/3 MPI Fortran                             OK       0.39 s

Ok:                    3
Expected Fail:         0
Fail:                  0
Unexpected Pass:       0
Skipped:               0
Timeout:               0
```

**Comment by jpakkane** [comment]
Are those regressions or were they broken already? If the latter, how would we get them fixed?

**Comment by scivision** [comment]
those would be regressions, the "failing" test cases above were working with prior Meson releases as I use  MPI with Meson regularly with those compilers / OS

**Comment by dcbaker** [comment]
I can poke at those with ICL later today and see if I can get that working.

**Comment by dcbaker** [comment]
I fixed icc (linux), ICL is *still* running updates so I haven't had a chance to check that yet.

**Comment by dcbaker** [comment]
@scivision, ping :)

**Pull request merged**